src/core/raster/qgsrasterdrawer.cpp

```diff
@@ -149,7 +152,7 @@ void QgsRasterDrawer::drawImage( QPainter *p, QgsRasterViewPort *viewPort, const
     }
   }
 
-  p->drawImage( tlPoint, img );
+  p->drawImage( tlPoint, dpiScaleFactor != 1.0 ? img.scaledToWidth( img.width() / dpiScaleFactor ) : img );
```

Comment:
Just wondering if you tried scaling the painter in the raster layer renderer instead of modifying QgsRasterDrawer like we discussed?